src/hypertable.c

```diff
@@ -1849,77 +1853,120 @@ ts_hypertable_create_internal(PG_FUNCTION_ARGS, bool is_dist_call)
+			data_nodes = ts_cm_functions->get_and_validate_data_node_list(data_node_arr);
 
-		/* If the number of partitions isn't specified, default to setting it
-		 * to the number of data nodes */
-		if (num_partitions < 1 && replication_factor > 0)
+		if (NULL != space_dim_name)
 		{
-			int num_nodes = list_length(data_nodes);
-
-			Assert(num_nodes >= 0);
-			num_partitions = num_nodes & 0xFFFF;
+			int16 num_partitions = PG_ARGISNULL(3) ? -1 : PG_GETARG_INT16(3);
+
+			/* If the number of partitions isn't specified, default to setting it
+			 * to the number of data nodes */
+			if (num_partitions < 1 && replication_factor > 0)
+			{
+				int num_nodes = list_length(data_nodes);
+
+				Assert(num_nodes >= 0);
+				num_partitions = num_nodes & 0xFFFF;
+			}
+
+			space_dim_info =
+				ts_dimension_info_create_closed(table_relid,
+												/* column name */
+												space_dim_name,
+												/* number partitions */
+												num_partitions,
+												/* partitioning func */
+												PG_ARGISNULL(9) ? InvalidOid : PG_GETARG_OID(9));
 		}
 
-		space_dim_info =
-			ts_dimension_info_create_closed(table_relid,
-											/* column name */
-											space_dim_name,
-											/* number partitions */
-											num_partitions,
-											/* partitioning func */
-											PG_ARGISNULL(9) ? InvalidOid : PG_GETARG_OID(9));
+		if (if_not_exists)
+			flags |= HYPERTABLE_CREATE_IF_NOT_EXISTS;
+		if (!create_default_indexes)
+			flags |= HYPERTABLE_CREATE_DISABLE_DEFAULT_INDEXES;
+		if (migrate_data)
+			flags |= HYPERTABLE_CREATE_MIGRATE_DATA;
+
+		created = ts_hypertable_create_from_info(table_relid,
+												 INVALID_HYPERTABLE_ID,
+												 flags,
+												 time_dim_info,
+												 space_dim_info,
+												 associated_schema_name,
+												 associated_table_prefix,
+												 &chunk_sizing_info,
+												 replication_factor,
+												 data_nodes);
+
+		ht = ts_hypertable_cache_get_cache_and_entry(table_relid, CACHE_FLAG_NONE, &hcache);
+
+		if (NULL != space_dim_info)
+		{
+			/* Function `ts_hypertable_create_from_info` will error if the
+			 * hypertable already exists unless the if-not-exists flag is not set.
+			 * If we reach this point, either if-not-exists-flag was set or the
+			 * hypertable did not exist and was created above.
+			 *
+			 * In `ts_hypertable_create_from_info`, a call to
+			 * `ts_dimension_info_validate(space_dim_info)` will be made if (and
+			 * only if) the table did not exist. The function does not only
+			 * validate the dimension, it also set `dimension_id` field.
+			 *
+			 * If the table already existed, `created` will be false and the
+			 * `dimension_id` will be set to the invalid OID, which means that
+			 * `ts_hypertable_check_partitioning` will crash since it it expect a
+			 * proper OID to be passed. This means that we have to fetch the
+			 * dimension OID here in the event that the table was not created. */
+			if (!created)
```

Comment:
I don't think this case is possible anymore, so this is dead code. Should `Assert(created)`

Reply:
I added the checks you showed above mine: if the code is dead, it will not make a difference to the test result.

If you look at the regression tests, you see the same failures that I got locally. It looks like I have applied the patch you showed correctly, but maybe I've made a mistake.

Reply:
In my version that I tested with I did something similar: I had the check a little further down. Since the following two checks just check input parameters, there is no point in looking up the hypertable at that time.

Reply:
I am not following. We are in top-level `else` clause, which means the hypertable did not exist, so `created` must be true (or there is an error). Checking `if (!created)` here makes no sense. These changes are, I believe, part of you previous fix prior to the early check for the hypertable so I don't understand why this is needed now given that I don't see a case for `created` being `false`.